### ISSUE TITLE
Keep ordering of deserialized maps

### ISSUE THREAD

**Issue by dlo9** [mehcode/config-rs#216]
Serde's yaml deserialization supports maintaining the order of map types (at least in yaml/json, haven't looked at other formats) by using `LinkedHashMap` instead of `HashMap`. Quickly perusing through `config-rs`, it looks like the `table` value uses `HashMap` and so will always lose the ordering of a map.

AFAICT, a fix to this would be as simple as changing `HashMap` to `LinkedHashMap` in the appropriate places, but since that type is exposed to the API, it could be considered a breaking change. Would the project be receptive to such a change?

**Comment by matthiasbeyer** [comment]
Hi,

thanks, that's a nice idea to pursue. Maybe we can work together and explore whether it would be possible to make the YAML de/serialization mechanisms generic over `HashMap` vs `LinkedHashMap`? I don't know whether such a thing is possible, but maybe it is...
If not, maybe adding a new "backend type" which explicitely uses `LinkedHashMap` is possible?

If all fails, we can always release breaking changes as long as we're in version `0.x.y`, although I would prefer to not have to do this.

Long story short: Consider me interested! :laughing:

**Comment by dlo9** [comment]
Great! I'll see if can can find some time in the next few days to make a test case and initial breaking implementation, and then we can go from there

**Comment by dlo9** [comment]
Just made a draft PR for an initial implementation -- there's a few things I discovered while making the change which I'd like to fix, but it will probably take me a few days to get around to. In the meantime, let me know if you have any changes or requested comments and I'll make it ready when I'm happy with the PR
